### PR TITLE
CellProfiler - ImageMath tool

### PR DESCRIPTION
**Pull request by beatrizserrano**
@sunyi000 @bgruening @davelopez please have no mercy 😄

### CODE REVIEW

File: tools/cellprofiler/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+.venv/
```
Comment (bgruening) [inline tools/cellprofiler/.gitignore:1]: this should be at the top level :)

File: tools/cellprofiler/cellprofiler.xml [hunk 2]
```diff
@@ -1,5 +1,5 @@
-<tool id="cp_cellprofiler" name="cellProfiler" version="@CP_VERSION@">
-    <description>runs a cellProfiler pipeline assembled by all the cellProfiler tools upstream in the workflow</description>
+<tool id="cp_cellprofiler" name="CellProfiler" version="@CP_VERSION@">
+    <description>run a CellProfiler pipeline assembled by all the CellProfiler tools upstream in the workflow</description>
```
Comment (bgruening) [inline tools/cellprofiler/cellprofiler.xml:2]: > by all the CellProfiler tools upstream in the workflow

Is that true, if you want to allow also uploading preconfigured pipelines?

File: tools/cellprofiler/cp_common_functions.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        value = input_params[key_list[0]]
+        for k in key_list[1:]:
+            value = value[k]
+        return(value)
+    except KeyError:
+        return("")
+
+
+def concat_conditional(a, b):
+    if (a == "" or b == ""):
```
Comment (bgruening) [inline tools/cellprofiler/cp_common_functions.py:19]: please remove brackets

File: tools/cellprofiler/cp_common_functions.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    if not isinstance(input_params, dict):
+        return ""
+    key_list = keys_path.split(".")
+    try:
+        value = input_params[key_list[0]]
+        for k in key_list[1:]:
+            value = value[k]
+        return(value)
+    except KeyError:
+        return("")
```
Comment (bgruening) [inline tools/cellprofiler/cp_common_functions.py:15]: remove brackets

File: tools/cellprofiler/cp_common_functions.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+def get_json_value(input_params, keys_path):
+    """Returns the value in keys_path or None if the field does not exist"""
+    if not isinstance(input_params, dict):
+        return ""
+    key_list = keys_path.split(".")
+    try:
+        value = input_params[key_list[0]]
+        for k in key_list[1:]:
+            value = value[k]
+        return(value)
```
Comment (bgruening) [inline tools/cellprofiler/cp_common_functions.py:13]: This looks like c# ;)

File: tools/cellprofiler/image_math.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    second_image_block = build_second_image_block(
+        get_json_value(inputs_galaxy, 'operation.second_image'))
+    result = "\n".join([result, second_image_block])
+    return (result)
+
+
+def build_first_image_block(input_params):
+    """Creates the block of parameters for the first operator in operations"""
+    result = INDENTATION.join(
+        [f"{INDENTATION}Image or measurement?:{get_json_value(input_params,'image_or_measurement_first.image_or_measurement_first')}\n",
```
Comment (bgruening) [inline tools/cellprofiler/image_math.py:59]: wouldn't it be better to take the get_json_value out of the string production, the function call directly before the `result =`?

File: tools/cellprofiler/image_math.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
+import json
```
Comment (bgruening) [inline tools/cellprofiler/image_math.py:1]: please add a shebang ... `#!/usr/bin/env python`

File: tools/cellprofiler/image_math.xml [hunk 8]
```diff
@@ -0,0 +1,301 @@
+<tool id="image_math" name="ImageMath" version="0.1.0" python_template_version="3.5">
```
Comment (bgruening) [inline tools/cellprofiler/image_math.xml:1]: we have started all other tool IDs with cp_ I think.

File: tools/cellprofiler/image_math.xml [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+    <description>perform simple mathematical operations on image intensities in CellProfiler</description>
```
Comment (bgruening) [inline tools/cellprofiler/image_math.xml:2]: in or with CP?

File: tools/cellprofiler/image_math.xml [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+    <macros>
+        <import>macros.xml</import>
+        <xml name="select_category" token_order="first">
```
Comment (bgruening) [inline tools/cellprofiler/image_math.xml:5]: this one is only used once? Is there no second-order?

File: tools/cellprofiler/image_math.xml [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+                    <option value="No">No</option>
+                </param>
+            </section>
+        </xml>
+    </macros>
+    <requirements>
+    </requirements>
+    <command detect_errors="exit_code">
+        <![CDATA[
+            python '$__tool_directory__/image_math.py' -p '$input_pipeline' -i '$inputs'
```
Comment (bgruening) [inline tools/cellprofiler/image_math.xml:120]: input could be named, like input_configfile or something like that ... maybe people expect here a image or so

File: tools/cellprofiler/image_math.xml [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+            <param name="name_output_image" value="ImageAfterMath"/>
+            <param name="select_the_first_image" value="DNA"/>
+            <param name="ignore_the_image_masks" value="No"/>
+            <output name="output_pipeline" ftype="txt" file="image_math_3.cppipe" lines_diff="0"/>
+        </test>
+    </tests>
+    <help>
+        <![CDATA[
+
+            .. class:: infomark
```
Comment (bgruening) [inline tools/cellprofiler/image_math.xml:289]: I don't think you can use the indentation here. Have you look with planmo serve how this looks like?

File: tools/cellprofiler/macros.xml [hunk 13]
```diff
@@ -2,72 +2,104 @@
...
+
+        **Input**
+
+        Existing CellProfiler pipeline file *(.cppipe)* or generated by linking CellProfiler tools.
+
+        .. class:: infomark
+
+        **Output**
+
+        The input CellProfiler pipeline file *(.cppipe)* including the settings of this module.
```
Comment (bgruening) [inline tools/cellprofiler/macros.xml:18]: maybe instead of "incuding", in addtition .... ?

File: tools/cellprofiler/macros.xml [hunk 14]
```diff
@@ -2,72 +2,104 @@
...
+
+        .. class:: warningmark
+
+        **IMPORTANT**
+
+        The first tool in a CellProfiler workflow has to be **Starting modules** and the last one **CellProfiler**.
```
Comment (bgruening) [inline tools/cellprofiler/macros.xml:24]: >  The first tool in a CellProfiler workflow has to be **Starting modules** and the last one **CellProfiler**.

 The first tool in a CellProfiler workflow has to be **Starting modules**. You execute the entire pipeline with the final **CellProfiler** tool, in which you feed in the images you want to process as well.

File: tools/cellprofiler/starting_modules.xml [hunk 15]
```diff
@@ -1,5 +1,5 @@
-<tool id="cp_common" name="cellProfiler common start modules" version="@CP_VERSION@">
-    <description>load images and metadata into cellProfiler</description>
+<tool id="cp_common" name="Starting modules" version="@CP_VERSION@">
+    <description> (Images, Metadata, NamesAndTypes, Groups) load images and metadata into CellProfiler</description>
```
Comment (bgruening) [inline tools/cellprofiler/starting_modules.xml:2]: the bracket at the end?

File: tools/cellprofiler/test-data/common_image_math.cppipe [hunk 16]
```diff
@@ -0,0 +1,94 @@
+CellProfiler Pipeline: http://www.cellprofiler.org
```
Comment (bgruening) [inline tools/cellprofiler/test-data/common_image_math.cppipe:1]: is this how a file always needs to look like?

File: tools/cellprofiler/test-data/common_image_math.cppipe [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+Version:3
```
Comment (bgruening) [inline tools/cellprofiler/test-data/common_image_math.cppipe:2]: can we estimate some compatibility things from those versions?

File: tools/cellprofiler/image_math.py [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+         f"Name the output image:{get_json_value(input_params,'name_output_image')}"
+         ])
+    return(result)
+
+
+def build_variable_block(inputs_galaxy):
+    result = ""
+    first_image_block = build_first_image_block(
+        get_json_value(inputs_galaxy, 'operation.first_image'))
+    result = "\n".join([result, first_image_block])
```
Comment (bgruening) [inline tools/cellprofiler/image_math.py:49]: Maybe that is easier in those simple cases? `result += "\n{first_image_block}"

File: tools/cellprofiler/image_math.py [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+    second_image_block = build_second_image_block(
+        get_json_value(inputs_galaxy, 'operation.second_image'))
+    result = "\n".join([result, second_image_block])
```
Comment (bgruening) [inline tools/cellprofiler/image_math.py:52]: or why not simply summing everything up only once and safe the first result step?
Please keep in mind you can do `a += b`

File: tools/cellprofiler/.gitignore (hunk 1 above)
Comment (beatrizserrano) [reply tools/cellprofiler/.gitignore:1]: included in the common .gitignore you mean?

Comment (bgruening) [reply tools/cellprofiler/.gitignore:1]: yes.

File: tools/cellprofiler/cellprofiler.xml (hunk 2 above)
Comment (beatrizserrano) [reply tools/cellprofiler/cellprofiler.xml]: changed the description

File: tools/cellprofiler/test-data/common_image_math.cppipe (hunk 16 above)
Comment (beatrizserrano) [reply tools/cellprofiler/test-data/common_image_math.cppipe:1]: I guess so... but not sure, maybe @sunyi000 knows :)

File: tools/cellprofiler/test-data/common_image_math.cppipe (hunk 17 above)
Comment (beatrizserrano) [reply tools/cellprofiler/test-data/common_image_math.cppipe:2]: not compatible with Version 4 due to the encoding problem :(

File: tools/cellprofiler/image_math.py (hunk 18 above)
Comment (beatrizserrano) [reply tools/cellprofiler/image_math.py:49]: consistency? ;)

File: tools/cellprofiler/cp_common_functions.py [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+def concat_conditional(a, b):
+    if (a == "" or b == ""):
+        return ""
+    else:
+        return f"{a}_{b}"
+
+        
+def get_total_number_of_modules(pipeline_lines):
+    """Gets the number of modules from the header of the previous pipeline"""
+    number_of_modules = pipeline_lines[4].strip().split(':')[1]
```
Comment (davelopez) [inline tools/cellprofiler/cp_common_functions.py:27]: Maybe put the magic number in some constant?

File: tools/cellprofiler/cp_common_functions.py [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+def get_pipeline_lines(input_pipeline):
+    """Returns a list with the lines in the .cppipe file"""
+    with open(input_pipeline) as f:
+        lines = f.readlines()
+    return(lines)
+
+
+def update_module_count(pipeline_lines, count):
+    """Updates the number of modules in the .cppipe header"""
+    module_count_entry = pipeline_lines[4].strip().split(':')[0]
```
Comment (davelopez) [inline tools/cellprofiler/cp_common_functions.py:40]: The magic number again :)

File: tools/cellprofiler/starting_modules.xml (hunk 15 above)
Comment (beatrizserrano) [reply tools/cellprofiler/starting_modules.xml]: did you mean here that the description should be shorter?

Comment (bgruening) [reply tools/cellprofiler/starting_modules.xml]: No, just "<description>load images and metadata into CellProfiler (Images, Metadata, NamesAndTypes, Groups)</description>"

File: tools/cellprofiler/cp_common_functions.py [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+    if not isinstance(input_params, dict):
+        return ""
+    key_list = keys_path.split(".")
+    try:
+        value = input_params[key_list[0]]
+        for k in key_list[1:]:
+            value = value[k]
+        return value
+    except KeyError:
+        return ""
```
Comment (davelopez) [inline tools/cellprofiler/cp_common_functions.py:15]:
```suggestion
def get_json_value(json_input, keys_path):
    """Returns the value specified in keys_path (using dot notation) or an empty string if the key doesn't exist"""
    if not isinstance(json_input, dict):
        return ""
    keys = keys_path.split(".")
    try:
        value = json_input[keys[0]]
        for key in keys[1:]:
            value = value[key]
        return value
    except KeyError:
        return ""
```

File: tools/cellprofiler/image_math.xml (hunk 10 above)
Comment (beatrizserrano) [reply tools/cellprofiler/image_math.xml:5]: used at least once but in most of the cases, twice. Is `token_order="first"` not appropriate here?

File: tools/cellprofiler/image_math.xml (hunk 12 above)
Comment (beatrizserrano) [reply tools/cellprofiler/image_math.xml:288]: looks good :)

**Pull request merged**